src/plugins/data/public/data_sources/datasource/factory.ts

```diff
@@ -0,0 +1,77 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+/**
+ * The DataSourceFactory is responsible for managing the registration and creation of data source classes.
+ * It serves as a registry for different data source types and provides a way to instantiate them.
+ */
+
+import { DataSourceType } from '../datasource_services';
+import { DataSource } from '../datasource';
+
+type DataSourceClass<
+  MetaData = any,
+  SetParams = any,
+  DataSet = any,
+  QueryParams = any,
+  QueryResult = any
+> = new (config: any) => DataSource<MetaData, SetParams, DataSet, QueryParams, QueryResult>;
+
+export class DataSourceFactory {
+  // Holds the singleton instance of the DataSourceFactory.
+  private static factory: DataSourceFactory;
+
+  // A dictionary holding the data source type as the key and its corresponding class constructor as the value.
+  private dataSourceClasses: { [type: string]: DataSourceClass } = {};
+
+  /**
+   * Private constructor to ensure only one instance of DataSourceFactory is created.
+   */
+  private constructor() {}
+
+  /**
+   * Returns the singleton instance of the DataSourceFactory. If it doesn't exist, it creates one.
+   *
+   * @returns {DataSourceFactory} The single instance of DataSourceFactory.
+   */
+  static getInstance(): DataSourceFactory {
+    if (!this.factory) {
+      this.factory = new DataSourceFactory();
+    }
+    return this.factory;
+  }
```

Comment:
For a singleton, all you should need is
```js
export const instance = new DataSourceFactory();
```
... outside the class.

Reply:
but it will always be initialized and take memeory. current way it's lazy init

Reply:
discussed offline and you mentioned the code will run at the time it is imported, so not always initialized in the beginning. i believe that would be in this file
https://github.com/opensearch-project/OpenSearch-Dashboards/blob/3642ef5938360b1c7d1b0d064b3d3b1f1a7b8b9a/src/plugins/data/public/plugin.ts#L220

But when i tested locally, exporting here is enough to run `const instance = new DataSourceFactory();`. i think it is created whenever data plugin is imported, so i still think it's better to initialize at `.getInstance()` call than initialize when importing the file